### DISCUSSION TITLE
Repository request for Python libraries for electrical matching networks

### DISCUSSION THREAD

**Discussion by thePlasm** [dtudk/Admin#4] [Ask for a sub-repo]
I am a PhD student in the Electromagnetic Systems (EMS) division in DTU Space. My PhD project involves the development of magnetic resonance (MR) radio frequency (RF) coils, hence I work with the MR section in the Healthtech department as well.

I am currently working on two papers relating to RF coil matching networks. I have written Python libraries/Jupyter notebooks to perform calculations for these matching networks. I am requesting for repositories to be made such that I can publish these libraries to accompany these papers. Once these repos have been published, I will use Zenodo to give them persistent identifiers.

The names of the repositories would be:
1. four-element (For "Four-Element LC-Baluns for Power Matching Arbitrary Impedances")
2. five-element (For "Five-Element Matching and Decoupling Networks for Radio-Frequency Receive Coils Providing Common-Mode Rejection")

Please let me know if these repo names are too generic. It would be appreciated if these repos could be created as private repos initially such that I can organise everything before making the repo public. These repositories will be licensed under the LGPLv2.1 as they use Lcapy which is licensed under the LGPLv2.1. My project is also partially ERC/EU funded, so I will add a statement for this to the README files. The repos could be created under a group for the EMS division or MR section, though it may be useful to create them under a group relating to MR coils or just directly under the dtudk namespace.

I have other RF coil related code that I would like to publish in the future, though I will advise when I will need to create new repos. This other code will depend on these repos, hence I will publish these two repos as PyPI packages so that they can be installed with pip.

Please let me know if any further information is needed.

Regards,
Rasmus Jepsen

**Comment by zerothi** [comment]
Thanks! I am a bit overburdened this week, I'll get back to you start of next week.

1. The names are to generic, I would advice you to incorporate some of the other words into the repo name. You should consider these names should also reflect the pypi packages. Say something like `rf4-lcbpmai` and `rf5-dnrfrcpcm` or whatever, feel free to make them more appropriate. 
2. We can start it private, that should be no problem

**Reply by thePlasm** [reply]
Would the following be more appropriate for the repo names?
1. lc-power-match-baluns
2. mr-recv-coil-match-net

**Reply by zerothi** [reply]
yes, much better :)
I would advice that your repo names are reflecting the pypi packages (for ease of use), but they _can_ be different. Would they be appropriate?

**Reply by thePlasm** [reply]
I was originally going to use four_element and five_element for the pypi packages, though I can change them to e.g. lc_power_match_baluns, as there isn't much code that depends on those names at the moment. It also seems like the names in my reply don't conflict with any existing pypi package names.

**Reply by zerothi** [reply]
great, IMHO having a bit of meaning would be favourable as four_element could cover many other fields, whereas your current names are probably more suitable :)

**Reply by zerothi** [reply]
It typically doesn't matter with the `-` vs `_` so you can choose here whatever you wish. :)

Please let me know your final decision of the repo names

**Reply by thePlasm** [reply]
Thanks for your reply. I would like to stick to lc-power-match-baluns and mr-recv-coil-match-networks for the repo names. I will replace the hyphens with underscores for the pypi packages to prevent issues with Python import.

**Reply by zerothi** [reply]
Ok, FYI, I have seen others stick with the `hyphen` on package install, and this would force users to use `_` in imports (obviously). Then the package repo and pip installation are similar, while the package usage is `_`. Just FYI, still your decision :)

**Reply by zerothi** [reply]
You should have access to the repo's now. They are currently private. I am not fully sure if you can publish them, if not, and when you want them public, let me know! :)

Thanks!

**Reply by thePlasm** [reply]
I see, if that is the norm, then I will use hyphens for the repo names and pypi packages, but underscores for the actual python imports.

Please let me know if you need any other information to create the repos.

**Reply by zerothi** [reply]
They have already been created, you can go ahead :)

**Reply by thePlasm** [reply]
Thanks for that.

**Reply by thePlasm** [reply]
Would it also be possible to grant me admin access to the new repositories such that I can adjust the default branch and other repo settings (e.g. repo description)?

**Reply by zerothi** [reply]
Sadly not, giving `admin` rights allows things we have to safe-keep...

But let me know, and I'll adjust the repo-description and default branch.

**Comment by thePlasm** [comment]
Dear Nick,

For the four-element paper, I would also like to publish Jupyter notebooks for processing experimental data and for the theoretical derivations of the baluns. Would it be preferred to have these notebooks in the same repo as the PyPI package for performing design calculations? Alternatively, would it be better to have these notebooks in separate repos (e.g. lc-power-match-baluns-experiment, lc-power-match-baluns-derivations)? Or just have the notebooks in a separate repo (e.g. lc-power-match-baluns-notebooks)?

Regards,
Rasmus

**Reply by zerothi** [reply]
I would put them in the same repo, say in an `examples` directory, or something similar. Just be careful about having executed code in the notebooks (only commit clean notebooks) :)

**Comment by thePlasm** [comment]
Dear Nick,

I have populated the `lc-power-match-baluns` repository with the code/data for an initial release. I have committed to a `draft` branch, which I will merge into a `main` branch when publishing. I have added code/data to the `experiment` directory, please let me know if should not have the Touchstone files and images in the repository.

Would it be possible for you to look at the repository to ensure that I haven't made any mistakes (with respect to e.g. copyright notices, README, folder/package structure, notebooks, etc.)? Once the repo has been published, I will add more detail to the README (e.g. regarding PyPI installation instructions, Zenodo integration and citation information).

I would also like to add more notebooks for the derivations of the balun design equations, though I might add these after the repo is published (possibly as a second release).

Please let me know if you have any other feedback regarding the repo.

Regards,
Rasmus

**Reply by zerothi** [reply]
I'll have a look next week (currently in autumn break).

**Reply by zerothi** [reply]
Hi Rasmus,

Some comments:

- I would recommend you to remove the `setup.py` it looks so simple that it isn't needed, 
Just have this in your `pyproject.toml` 
```
[build-system]
requires = [
    "setuptools_scm[toml]>=6.2",
]
```
- Correct the other links in `pyproject.toml`
- Put the `lcapy` and Python requirement in the `pyproject.toml` as well, something like this:
```
[project]
requires-python = ">=3.10" # is 3.10 really a hard limit? Are you using functionalities from 10, for instance whalrus operator?
dependencies = ["lcapy==1.15"] # is this also a hard limit? would lcapy~=1.15 equally work
```
- when having the dependencies in the `pyproject.toml` users don't need to do the `pip install -r requirements.txt` step, they can just do `pip install .`
- You could also add to the documentation how to install without downloading the details
- you are using a simple python project, is `hatchling` really necessary? Wouldn't setuptools be sufficient, it is more accessible and doesn't require a big deal.
- the `experiments` folder. It is your decision, I tend to separate things that are not strictly related. For instance, are they only there for a documentation purpose, if so then I would probably remove them... I would advice on you creating a `MANIFEST.in` to ensure that the `experiment` folder isn't added to the wheel. I.e. if you plan on publishing the package on pypi, then the wheel would contain all the things, unless you manually control this via `MANIFEST.in`.

These are just a first round of comments, hope they'll guide you initially.

**Reply by zerothi** [reply]
sorry, I mistyped in the first requires block, see the updated code

**Reply by thePlasm** [reply]
Thanks for your feedback. I am a bit pre-occupied this week, though I'll have a look into this next week.

**Reply by thePlasm** [reply]
Thanks again for your feedback. I have updated the repo to address these points:
* `setup.py` and `requirements.txt` have been removed and `pyproject.toml` has been updated accordingly
* I am not sure what you mean by "You could also add to the documentation how to install without downloading the details", though I have updated the documentation to remove the part about `requirements.txt` and to fix some attribution details
* I have added a `MANIFEST.in` file to exclude the `experiment` and future `derivations` folders
* I have also changed the `@staticmethod`s to `@classmethod`s as per your other issue

Please let me know if there are any further issues that I should fix.

I will work on finalising the experimental data and derivation notebooks for this repo over the next week and a bit. Following which, I will setup publishing with PyPI. Once this is done, the repo should be ready to publish publicly on GitHub and Zenodo.

**Reply by zerothi** [reply]
Great, as for point 2, you can install directly from a git repo using:
```
python3 -m pip install git+https://github.com/dtudk/<repo>sisl.git
```
which is simpler for end-users than first downloading, then installing. Just for you to know.

I think all is fine.

**Comment by zerothi** [comment]
I think everything is solved, I'll close this discussion.
